falcon/util/misc.py

```diff
@@ -305,3 +308,25 @@ def get_http_status(status_code, default_reason='Unknown'):
     except AttributeError:
         # not found
         return str(code) + ' ' + default_reason
+
+
+def secure_filename(filename):
+    """
+    Sanitize the provided `filename` using only the most common ASCII
+    characters for maximum portability and safety wrt using this name as a
+    filename on a regular file system.
+
+    Args:
+        filename (str): Arbitrary filename input from the request, such as
+            multipart form filename field.
+
+    Returns:
+        str: sanitized filename
+    """
+    # TODO(vytas): max_length (int): Maximum length of the returned
+    #     filename. Should the returned filename exceed this restriction, it is
+    #     truncated while attempting to preserve the extension.
+    filename = unicodedata.normalize('NFKD', filename)
+    if filename.startswith('.'):
+        filename = filename.replace('.', '_', 1)
+    return _UNSAFE_CHARS.sub('_', filename) or 'file'
```

Comment:
I was unsure about this whole function, and the corresponding `BodyPart.secure_filename` property, whether this is needed in Falcon at all. OTOH Werkzeug has it, and a property is a low overhead porcelain.

Agreed about that `or 'file'`. We could:
*  Just note in the docstring that this function may return an empty string
* Add a default kwarg parameter, like `fallback_filename='something'`. OTOH the property won't be able to utilize it directly, and adding this to parser options might be a bit overboard
* Keep 'or 'file'` with a more descriptive default name, suggestions welcome for that name :slightly_smiling_face:

Reply:
I vote for returning an empty string and documenting that. It's easy enough for someone to key off of that and use their own default logic in that case or raise an error. If they don't have any handling for it, when they go to save it somewhere the thing they are using should refuse to write the data...which is actually IMHO what you want since it means you got some likely malicious input. Actually, with that in mind, maybe we should actually be raising an error in this case rather than an empty string or default name?

Reply:
I went on to implement the exception path, thanks!